### ISSUE TITLE
Invalid actions in a custom ActorDistributionNetwork for a DdpgAgent

### ISSUE THREAD

**Issue by JCMiles** [tensorflow/agents#50]
Hello team, I will try to generalize my question as much as I can, so maybe it could be useful also for other users. The simplified version of my custom actor network looks like this:

@gin.configurable
class Actor(DistributionNetwork):
    def __init__(self, spec, name='ActorNetwork'):
        super(Actor, self).__init__(
            input_tensor_spec=spec.input_tensor_spec,
            state_spec=spec.state_spec,
            output_spec=spec.output_spec,
            name=name)

        self._forward.add(layers.Dense(50))
        self._forward.add(layers.Dense(10))

    def call(self, observation, step_type, network_state=None):   
        logits = self._forward(observation)
        return logits, network_state

ENVIRONMENT DEFINITION:
The Environment is a continuos action space and at each timestep
my state is represented by N features + 1 binary feature, let's call this last one "transition". So by definition the transition feature can be only 0 or 1 and every time we reset the environment the sequence always looks like this:
0101010101010101010101 .....

At each timestep only 5 actions are allowed for the agent,
let's say:
transition = 0   ---> valid actions are logits[:5]
transition = 1   ---> valid actions are logits[5:]

In my previous code in tf.1 I was handling the illegal actions
simply by filtering them in this way:

def valid_actions(logits, transition):
    return tf.cond(tf.equal(transition, 0),
                   lambda: tf.slice(logits, [0], [5]),
                   lambda: tf.slice(logits, [5], [5]))
                         
so the input for my critic network _action_layers is:
a = valid_actions(logits, transition)

Now my question,
it is correct to implement something like the "valid_actions function" above
directly at the end of the self.forward() pass ?
Or tf_agents provide some method to handle something like this specific case and even in  environements with more complicated structure of valid actions to handle ?

**Comment by kbanoop** [comment]
Yes using valid_actions at the end of `call` to filter logits sounds should work. Assuming `transition` is part of `observation`. 

We don't have a single generic way of handling structure in valid actions. In your case, what you suggested seems best to me. In some other cases, valid actions might depend on previously chosen actions, so we can store previously taken actions in the network/policy state and use that to decide the current action etc.
